### ISSUE TITLE
Support inlay hints

### ISSUE THREAD

**Issue by kitsonk** [denoland/vscode_deno#509]
TypeScript 4.4 which will be in Deno 1.14 supports inlay hints: https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#inlay-hints

We need to plumb that through in the LSP and integrate it into the extension.

**Comment by kitsonk** [comment]
Sadly, `InlayHints` are "experimental" in vscode at the moment, and there is no way to publish an extension that uses them, until the APIs are moved to stable.  Even then, they are currently not part of the language server protocol.  That isn't specifically a limitation as we can likely add the feature as a custom extension until it gets added.

**Comment by kitsonk** [comment]
Now vscode has shipped, but the API is still _proposed_, meaning we can't publish an extension that uses it.

Issue requesting it be made available to other extension authors: https://github.com/microsoft/vscode/issues/132197
There was an already existing issue to add it to the language server protocol, though it doesn't look like there was any movement of adding the vscode implementation to the specification ATM: https://github.com/microsoft/language-server-protocol/issues/956

**Comment by grian32** [comment]
It might be a good thing to add support for inlay hints to Deno LSP and not add it to the VSCode extension, Deno LSP support is on the [WebStorm 2021.3 Roadmap](https://blog.jetbrains.com/webstorm/2021/09/webstorm-2021-3-roadmap/) and WebStorm has support for inlay hints already.

**Comment by MihaelBercic** [comment]
> It is blocked by it being added to the LSP specification.

What is the solution to be able to use inlay hints with deno in the end?

**Comment by sno2** [comment]
The next LSP specification version will [include inlay hints](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlayHint).

**Comment by kitsonk** [comment]
The new APIs need to be added to (`lsp-types`)[https://github.com/gluon-lang/lsp-types] and then the changes need to be integrated into [`lspower`](https://github.com/silvanshade/lspower).  Once that is done, it should be possible to make them available in the lsp.  I would get around to doing that eventually, but it is a good opportunity for community contribution, and fairly straight forward changes.  Also contributions to `lsp-types` and `lspower` benefit a lot of Rust community language server implementations, not just Deno.

**Comment by kitsonk** [comment]
ah, ok... I should have paid closer attention... anyways, it needs to be integrated into `tower-lsp` then!

**Comment by kitsonk** [comment]
Closed by denoland/deno#16287, #739 and #740
